### PR TITLE
SDK 3.39.0: Adds version bump and changelog.

### PR DESCRIPTION
**Pull request by philipthomas-MSFT**
# Pull Request Template

## Description

This bumps the SDK GA version from 3.38.1 to 3.39.0 and preview version from 3.39.0-preview.1 to 3.40.0-preview.0. Please see the below changes in the GA public contract.

diff API_3_39_0

Also highlighted below are the preview contract changes.

diff API_3 40 0-preview 0

_[Note: To understand about the GA and preview versions styles, please take a look at our [versioning guidelines and principles](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/docs/versioning.md)]_

### CODE REVIEW

File: Directory.Build.props [hunk 1]
```diff
@@ -1,8 +1,8 @@
 <Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
 	<PropertyGroup>
-		<ClientOfficialVersion>3.38.1</ClientOfficialVersion>
-		<ClientPreviewVersion>3.39.0</ClientPreviewVersion>
-		<ClientPreviewSuffixVersion>preview.1</ClientPreviewSuffixVersion>
+		<ClientOfficialVersion>3.38.2</ClientOfficialVersion>
```
Comment (ealsur) [inline Directory.Build.props:3]: SHouldn't this be 3.39.0?

3.38.2 would be a hotfix release, not a full release from master. Hotfix releases cannot alter public surface (add new APIs): https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/docs/versioning.md

**Comment by kirankumarkolli** [comment]
Are contracts changes upto date? 

The images above shows some contracts/APIs are deleted right? Is that expected?

File: Directory.Build.props (hunk 1 above)
Comment (philipthomas-MSFT) [reply Directory.Build.props:3]: Thanks. I am fixing.

**Comment by philipthomas-MSFT** [comment]
> Are contracts changes upto date?
> 
> The images above shows some contracts/APIs are deleted right? Is that expected?

I should have checked this. Thanks for pointing it out.

**Comment by Maya-Painter** [comment]
The changelog doesn't seem to include any query changes - is this intentional?

**Comment by philipthomas-MSFT** [comment]
> The changelog doesn't seem to include any query changes - is this intentional?

Query PRs have been added. Thanks.

File: changelog.md [hunk 2]
```diff
@@ -13,6 +13,34 @@ Preview features are treated as a separate branch and will not be included in th
...
 
+### <a name="3.40.0-preview.0"/> [3.40.0-preview.0](https://www.nuget.org/packages/Microsoft.Azure.Cosmos/3.40.0-preview.0) - 2024-04-05
+
+#### Fixed
+- [4334](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4334) ChangeFeedProcessor: Fixes when ChangeFeedMode is switched, an exception is thrown
+- [4352](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4352) Documentation: Fixes spelling in ReplicaValidationDesign
+
+#### Added
+- [4370](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4370) ChangeFeedProcessor: Adds AllVersionsAndDeletes support to ChangeFeedProcessor 
+- [4380](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4380) ChangeFeedProcessor: Refactors AVAD Metadata Contract for ChangeFeed and ChangeFeedProcessor
```
Comment (kirankumarkolli) [inline changelog.md:24]: For preview please call out the breaking change very explicitly. 

This needs to pop-up differently from Added/fixed sections

File: changelog.md [hunk 3]
```diff
@@ -13,6 +13,34 @@ Preview features are treated as a separate branch and will not be included in th
...
+
+### <a name="3.39.0"/> [3.39.0](https://www.nuget.org/packages/Microsoft.Azure.Cosmos/3.39.0) - 2024-04-05
+
+#### Fixed
+- [4357](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4357) Distributed Tracing Documentation : Fixes the default value mentioned in code doc 
+
+#### Added
+- [4377](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4377) integrated cache: Adds BypassIntegratedCache for public release
+- [4281](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4281) FaultInjection: Adds Fault Injection Library
```
Comment (kirankumarkolli) [inline changelog.md:33]: Its not a end CX feature, ideally the PR should have marked it [INTERNAL]

Can we please exclude it?

File: changelog.md [hunk 4]
```diff
@@ -13,6 +13,34 @@ Preview features are treated as a separate branch and will not be included in th
...
+- [4265](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4265) CosmosClientOptions: Adds Private Custom Account Endpoints
+- [4316](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4316) Distributed Tracing: Refactors code to rename net.peer.name attribute to server.address
```
Comment (kirankumarkolli) [inline changelog.md:35]: @jcocchi should this be called as breaking change?

File: changelog.md [hunk 5]
```diff
@@ -13,6 +13,34 @@ Preview features are treated as a separate branch and will not be included in th
...
+- [4317](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4317) Client Telemetry: Refactors code to stop throwing an exception
```
Comment (kirankumarkolli) [inline changelog.md:36]: @sourabh1007 does it need to go into release notes?

Seems missing [INTERNAL] case.

File: changelog.md (hunk 3 above)
Comment (philipthomas-MSFT) [reply changelog.md]: removed. Thank you.

File: changelog.md (hunk 4 above)
Comment (jcocchi) [reply changelog.md:35]: Yes, the rename is breaking, but OTel attributes aren't stable yet. We should still call out this change

File: changelog.md [hunk 6]
```diff
@@ -13,6 +13,34 @@ Preview features are treated as a separate branch and will not be included in th
...
+- [4380](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4380) ChangeFeedProcessor: Refactors AVAD Metadata Contract for ChangeFeed and ChangeFeedProcessor
+  > Note: A Rename refactoring was performed in the effort to reduce redundancy and achieve clarity from a user perspective. The previous type ChangeFeedItemChange<T> was strategically renamed to ChangeFeedItem<T>. The refactoring affects both ChangeFeed (pull), and the new ChangeFeedProcessor (push), when in AllVersionsAndDeletes ChangeFeedMode. LatestVersion ChangeFeedMode is not affect and will continue to function as expected.
+
+### <a name="3.39.0"/> [3.39.0](https://www.nuget.org/packages/Microsoft.Azure.Cosmos/3.39.0) - 2024-04-05
+
+#### Fixed
+- [4357](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4357) Distributed Tracing Documentation : Fixes the default value mentioned in code doc 
+
+#### Added
+- [4377](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4377) integrated cache: Adds BypassIntegratedCache for public release
```
Comment (jcocchi) [inline changelog.md:33]: Nit for casing consistency

```suggestion
- [4377](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4377) Integrated Cache: Adds BypassIntegratedCache for public release
```

**Review by adityasa — DISMISSED** [review]
:shipit:

File: changelog.md (hunk 4 above)
Comment (philipthomas-MSFT) [reply changelog.md:35]: > Yes, the rename is breaking, but OTel attributes aren't stable yet. We should still call out this change

@sourabh1007 @jcocchi Can you ping me offline with what you want this to say, or just make the change and commit. Thanks.

**Review by NaluTripician — CHANGES_REQUESTED** [review]
Would probably be good to include that the new Taiwan regions are included in `Regions.cs` with the Direct Package upgrade in the changelog

**Comment by philipthomas-MSFT** [comment]
> Would probably be good to include that the new Taiwan regions are included in `Regions.cs` with the Direct Package upgrade in the changelog

I can agree with that. The PR was Internal. Was that a mistake?
#4386

File: changelog.md (hunk 4 above)
Comment (jcocchi) [reply changelog.md:35]: How about the below

```suggestion
- [4316](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4316) Distributed Tracing: Refactors code to rename net.peer.name attribute to server.address. **Warning:** This is a breaking change, only `server.address` will be emitted starting with this version.
```

**Comment by NaluTripician** [comment]
> > Would probably be good to include that the new Taiwan regions are included in `Regions.cs` with the Direct Package upgrade in the changelog
> 
> I can agree with that. The PR was Internal. Was that a mistake? #4386

I think normally Direct package upgrades are marked as Internal, but for this upgrade I was unaware of the region changes and that is why it was marked that way. I think we can add something like [this ](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/4b2196bc6023dca0e6a396612efb4a1ca14cf09a/changelog.md?plain=1#L57)to the changelog and link to the direct upgrade PR. Also worth changing our direct upgrade guide to specify this which I will do.

File: changelog.md (hunk 5 above)
Comment (sourabh1007) [reply changelog.md:36]: Yes, it should be "internal", Client telemetry one.

Comment (philipthomas-MSFT) [reply changelog.md]: removing from changelog

**Comment by philipthomas-MSFT** [comment]
> > > Would probably be good to include that the new Taiwan regions are included in `Regions.cs` with the Direct Package upgrade in the changelog
> > 
> > 
> > I can agree with that. The PR was Internal. Was that a mistake? #4386
> 
> I think normally Direct package upgrades are marked as Internal, but for this upgrade I was unaware of the region changes and that is why it was marked that way. I think we can add something like [this ](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/4b2196bc6023dca0e6a396612efb4a1ca14cf09a/changelog.md?plain=1#L57)to the changelog and link to the direct upgrade PR. Also worth changing our direct upgrade guide to specify this which I will do.

Ok. @NaluTripician , added the new regions and direct package version update to the changelog. @kirankumarkolli

File: changelog.md [hunk 7]
```diff
@@ -13,6 +13,34 @@ Preview features are treated as a separate branch and will not be included in th
...
+- [4281](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4281) FaultInjection: Adds Fault Injection Library
+- [4265](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4265) CosmosClientOptions: Adds Private Custom Account Endpoints
+- [4316](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4316) Distributed Tracing: Refactors code to rename net.peer.name attribute to server.address
+- [4317](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4317) Client Telemetry: Refactors code to stop throwing an exception
+- [4339](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4338) Diagnostics: Adds Client Configuration for Synchronization context cases
+- [4333](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4333) Distributed Tracing: Adds configuration to disable network level tracing in sdk permanently
+- [4323](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4323) Query: Adds Support for LINQ Custom Serializer in Public Release
```
[the remark this answers is not in the document]
Comment (philipthomas-MSFT) [reply changelog.md:38]: @adityasa has already approved this PR once. Can you approve it again please?

**Review by NaluTripician — DISMISSED** [review]

**Review by sourabh1007 — APPROVED** [review]

**Review by kirankumarkolli — APPROVED** [review]

**Comment by kirankumarkolli** [comment]
@ealsur can you please also take a look?

File: Microsoft.Azure.Cosmos/contracts/API_3.38.2.txt [hunk 8]
```diff
@@ -0,0 +1,1576 @@
+namespace Microsoft.Azure.Cosmos
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/contracts/API_3.38.2.txt:1]: Remove this file

File: changelog.md [hunk 9]
```diff
@@ -13,6 +13,37 @@ Preview features are treated as a separate branch and will not be included in th
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+### <a name="3.40.0-preview.0"/> [3.40.0-preview.0](https://www.nuget.org/packages/Microsoft.Azure.Cosmos/3.40.0-preview.0) - 2024-04-05
+
+#### Fixed
+- [4334](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4334) ChangeFeedProcessor: Fixes when ChangeFeedMode is switched, an exception is thrown
+- [4352](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4352) Documentation: Fixes spelling in ReplicaValidationDesign
```
Comment (ealsur) [inline changelog.md:20]: This is a PR without any code changes that fixes a spelling on one of our repo docs, I would not include it. It's not anything shipped in the sdk

File: changelog.md [hunk 10]
```diff
@@ -13,6 +13,37 @@ Preview features are treated as a separate branch and will not be included in th
...
+
+#### Added
+- [4370](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4370) ChangeFeedProcessor: Adds AllVersionsAndDeletes support to ChangeFeedProcessor 
+- [4380](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4380) ChangeFeedProcessor: Refactors AVAD Metadata Contract for ChangeFeed and ChangeFeedProcessor
```
Comment (ealsur) [inline changelog.md:24]: Above item says AllVersionsAndDeletes and this one says AVAD, should we be consistent with the text for customer's clarity?

File: changelog.md [hunk 11]
```diff
@@ -13,6 +13,37 @@ Preview features are treated as a separate branch and will not be included in th
...
+- [4339](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4338) Diagnostics: Adds Client Configuration for Synchronization context cases
+- [4333](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4333) Distributed Tracing: Adds configuration to disable network level tracing in sdk permanently
+- [4323](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4323) Query: Adds Support for LINQ Custom Serializer in Public Release
+- [4362](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4362) Query: Adds support for non streaming ORDER BY
+- [4074](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4074) Query: Adds translation support for single key single value select GROUP BY LINQ queries
+- [4361](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4361) Performance: Refactors query prefetch mechanism
+- [4386](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4386) Regions: Adds new Regions
+  > Note: There is added support for the following regions: `USSec West Central`, `Taiwan North` and `Taiwan Northwest`.
```
Comment (ealsur) [inline changelog.md:43]: `USSec West Central` is internal, not public.

File: Microsoft.Azure.Cosmos/contracts/API_3.38.2.txt (hunk 8 above)
Comment (philipthomas-MSFT) [reply Microsoft.Azure.Cosmos/contracts/API_3.38.2.txt]: done.

File: changelog.md (hunk 9 above)
Comment (philipthomas-MSFT) [reply changelog.md]: done

File: changelog.md (hunk 11 above)
Comment (philipthomas-MSFT) [reply changelog.md]: done

File: changelog.md (hunk 10 above)
Comment (philipthomas-MSFT) [reply changelog.md]: done

**Review by ealsur — APPROVED** [review]

**Review by kirankumarkolli — APPROVED** [review]

**Pull request merged**